salt/map.jinja

```diff
@@ -1,240 +1,24 @@
-      'pygit2': 'py27-pygit2',
-      'config_path': '/usr/local/etc/salt',
-      'minion_service': 'salt_minion',
-      'master_service': 'salt_master',
-      'api_service': 'salt_api',
-      'syndic_service': 'salt_syndic',
-    },
-    'OpenBSD': {
-      'salt_master': 'salt',
-      'salt_minion': 'salt',
-      'salt_syndic': 'salt',
-      'salt_cloud': 'salt',
-      'salt_api': 'salt',
-      'salt_ssh': 'salt',
-      'config_path': '/etc/salt',
-      'minion_service': 'salt_minion',
-      'master_service': 'salt_master',
-      'python_git': 'py-GitPython',
-    },
-    'Windows': {
-      'salt_minion': 'salt-minion',
-      'config_path': 'C:\salt\conf',
-      'minion_service': 'salt-minion',
-    },
-    'MacOS': {
-      'salt_minion': 'com.saltstack.salt',
-      'salt_minion_pkg_source': '',
-      'salt_minion_pkg_hash': '',
-      'config_path': '/private/etc/salt',
-      'minion_service': 'com.saltstack.salt.minion',
-    },
-  }, merge=salt['grains.filter_by']({
-    'Ubuntu':  {
-      'pkgrepo': 'deb http://repo.saltstack.com/apt/' +
-      salt['grains.get']('os')|lower + '/' + osrelease + '/amd64/' + salt_release  + ' ' + salt['grains.get']('oscodename') + ' main',
-      'key_url': 'https://repo.saltstack.com/apt/' + salt['grains.get']('os')|lower + '/' + osrelease + '/amd64/' + salt_release + '/SALTSTACK-GPG-KEY.pub',
-      'pygit2': 'python-pygit2',
-      'gitfs': {
-        'pygit2': {
-          'install_from_source': False,
-          'git': {
-            'require_state': False,
-            'install_from_package': None,
-          },
-        },
-      },
-    },
-    'Raspbian': {
-      'pkgrepo': 'deb http://repo.saltstack.com/apt/' +
-      salt['grains.get']('os_family')|lower + '/' + salt['grains.get']('osmajorrelease', osrelease)|string + '/armhf/' + salt_release + ' ' + salt['grains.get']('oscodename') + ' main',
-      'key_url': 'https://repo.saltstack.com/apt/' + salt['grains.get']('os_family')|lower + '/' + salt['grains.get']('osmajorrelease', osrelease)|string + '/armhf/' + salt_release + '/SALTSTACK-GPG-KEY.pub',
-    },
-    'SmartOS': {
-      'salt_master': 'salt',
-      'salt_minion': 'salt',
-      'salt_syndic': 'salt',
-      'salt_cloud': 'salt',
-      'salt_api': 'salt',
-      'salt_ssh': 'salt',
-      'minion_service': 'salt:minion',
-      'master_service': 'salt:master',
-      'api_service': 'salt:api',
-      'python_dulwich': 'py27-dulwich',
-      'gitfs': {
-        'dulwich': {
-          'install_from_source': False,
-        }
-      },
-      'config_path': '/opt/local/etc/salt',
-      'master': {
-        'gitfs_provider': 'dulwich'
-      },
-    }
-  }, grain='os', merge=salt['pillar.get']('salt:lookup')))
-%}
-
-{## Merge the flavor_map to the default settings ##}
-{% do deep_merge(default_settings.salt,os_family_map) %}
+{# merge the osmap #}
+{% set osmap = salt['grains.filter_by'](osmap, grain='os') or{} %}
+{% do defaults.salt.update(osmap) %}
```

Comment:
This is updating instead of merging, so it's only giving the Ubuntu `gitfs` key, rather than merging it into the Debian `gitfs` key, leading to the loss of values shown [in the diff](https://github.com/saltstack-formulas/salt-formula/pull/395#issuecomment-457748633).

Reply:
The bare bones of a solution outlined.

```jinja
{# merge the osfamilymap #}
{% set osfamily = salt['grains.filter_by'](osfamilymap, grain='os_family') or{} %}
{% do defaults.salt.update(osfamily) %}

{# merge the osmap #}
{% set osmap = salt['grains.filter_by'](osmap, grain='os') or{} %}
{% do defaults.salt.update(osmap) %}
```

Replaced by:

```jinja
{% set osmappings = salt['grains.filter_by'](
    osfamilymap,
    grain='os_family',
    merge=salt['grains.filter_by'](
        osmap,
        grain='os',
    ),
) %}
{% do defaults.salt.update(osmappings) %}
```

Reply:
I was hoping to do the merge on the end, only one time, but I can change it to merge it everytime...

please see last commit

Reply:
`salt['defaults.merge']` -- now this is what I've been looking for.  Thanks!

Reply:
Yes, found it after seeing your commit: https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.defaults.html#salt.modules.defaults.merge.